File: Scoper.py





class Scope:
    def __init__(self, parent, expression, content, indentation=0):          # parent refers to its parent Scope
        self.expression = expression
        self.content = content                    # content is always made up of other Scopes
        self.parent = parent
        self.indentation = indentation
    def __str__(self):
        ret = self.indentation * ' ' + str(self.expression) + '\n'
        if len(self.content) > 0:
            ret += '\n'.join([str(scope) for scope in self.content])
            ret += '\n'
        return ret



'''
Takes a list of ExpressionWithIndentations.
Every Expression becomes a Scope, even if it has nothing underneath it (case in which content is [] for it)
Returns a wrapper Scope, containing a recursively scoped list of Scope
'''

def scopify(expressionsWithIndentation):
    wrapperScope = Scope(parent=None, expression=None, content=[], indentation=-1)
    previousScope = wrapperScope

    for expr in expressionsWithIndentation:
        
        if expr.indentation < previousScope.indentation:
            while expr.indentation < previousScope.indentation:
                previousScope = previousScope.parent
        
        if expr.indentation > previousScope.indentation:
            scope = Scope(
                parent = previousScope,
                expression = expr.expression,
                content = [],
                indentation = expr.indentation)
            previousScope.content.append(scope)
            previousScope = scope

        elif expr.indentation == previousScope.indentation:
            parent = previousScope.parent
            scope = Scope(
                parent = previousScope.parent,
                expression = expr.expression,
                content = [],
                indentation = expr.indentation)
            parent.content.append(scope)
            previousScope = scope

        else:
            print("This should not have happened")
        
    return wrapperScope

File: JS/Scoper.js

Utils = require('./Utils')

class Scope {
    constructor(parent, expression, content, indentation=0) {
        this.expression = expression
        this.content = content
        this.parent = parent
        this.indentation = indentation
    }

    toString() {
        let ret = Utils.spaces(this.indentation)
        if(this.expression != null) {
            ret += this.expression.toString()
        }
        ret += '\n'
        if (this.content.length > 0) {
            ret += this.content.map( scope => scope.toString() ).join('\n')
            ret += '\n'
        }
        return ret
    }

    toJsonObject() {
        return {
            expression : this.expression,
            content : this.content.map( scope => scope != null? scope.toJsonObject(): scope),
        }
    }
}




function scopify(expressionsWithIndentation) {
    let wrapperScope = new Scope(null, null, [], -1)
    let previousScope = wrapperScope

    for (let expr of expressionsWithIndentation) {

        if (expr.indentation < previousScope.indentation) {
            while (expr.indentation < previousScope.indentation) {
                previousScope = previousScope.parent
            }
        }

        if (expr.indentation > previousScope.indentation) {
            let scope = new Scope(previousScope, expr.expression, [], expr.indentation)
            previousScope.content.push(scope)
            previousScope = scope
        } else if (expr.indentation == previousScope.indentation) {
            let parent = previousScope.parent
            let scope = new Scope(parent, expr.expression, [], expr.indentation)
            parent.content.push(scope)
            previousScope = scope
        } else {
            console.log('Scope Error: this should not have happened')
        }

    }
    return wrapperScope
}


module.exports = {
    Scope : Scope,
    scopify : scopify
}

__requirer['Scoper'] = module.exports
__requirer['./Scoper'] = module.exports
